fdbclient/include/fdbclient/BlobCipher.h

```diff
@@ -181,6 +192,186 @@ struct hash<BlobCipherDetails> {
+	// Serializable fields
+
+	// Text cipher encryption information
+	BlobCipherDetails cipherTextDetails;
+	// Initialization vector
+	uint8_t iv[AES_256_IV_LENGTH];
+
+	AesCtrNoAuthV1() {}
+	AesCtrNoAuthV1(const BlobCipherDetails& textDetails, const uint8_t* ivBuf, const int ivLen, Arena& arena)
+	  : cipherTextDetails(textDetails) {
+		ASSERT_EQ(ivLen, AES_256_IV_LENGTH);
+		memcpy(&iv[0], ivBuf, ivLen);
+	}
+
+	bool operator==(const AesCtrNoAuthV1& o) const {
+		return cipherTextDetails == o.cipherTextDetails && memcmp(&iv[0], &o.iv[0], AES_256_IV_LENGTH) == 0;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		ar.serializeBytes(this, sizeof(AesCtrNoAuthV1));
+	}
+};
+
+struct ParsedEncryptHeaderDetails {
+	EncryptCipherMode encryptMode;
+	EncryptAuthTokenMode authMode;
+	EncryptAuthTokenAlgo authAlgo;
+	BlobCipherDetails textCipherDetails;
+	BlobCipherDetails headerCipherDetails;
+	uint8_t iv[AES_256_IV_LENGTH];
+};
+
+struct BlobCipherEncryptHeaderRef {
+	// Serializable fields
+
+	// HeaderFlags version tracker
+	uint16_t flagsVersion;
+	// Encryption algorithm header version tracker
+	uint16_t algoHeaderVersion;
+
+	// The on-disk format doesn't store std::variant, currently "serializer" doesn't support std::variant, the
+	// (de)serialization code serializes the relevant BlobCipherEncryptHeader and AlgoHeader structs for a given
+	// 'flagVersion' and 'algoHeaderVersion'. Refer BlobCipherEncryptHeaderRef::serialize() for more details.
+	std::variant<BlobCipherEncryptHeaderFlagsV1> flags;
+	std::variant<AesCtrNoAuthV1, AesCtrWithAuthV1<AUTH_TOKEN_HMAC_SHA_SIZE>, AesCtrWithAuthV1<AUTH_TOKEN_AES_CMAC_SIZE>>
+	    algoHeader;
+
+	BlobCipherEncryptHeaderRef()
+	  : flagsVersion(INVALID_ENCRYPT_HEADERS_FLAG_VERSION),
+	    algoHeaderVersion(INVALID_ENCRYPT_HEADER_ALGO_HEADER_VERSION) {}
+	BlobCipherEncryptHeaderRef(const BlobCipherEncryptHeaderRef& src)
+	  : flagsVersion(src.flagsVersion), algoHeaderVersion(src.algoHeaderVersion), flags(src.flags),
+	    algoHeader(src.algoHeader) {}
+
+	static BlobCipherEncryptHeaderRef fromStringRef(const StringRef& header) {
+		return BinaryReader::fromStringRef<BlobCipherEncryptHeaderRef>(header, IncludeVersion());
+	}
+
+	static Standalone<StringRef> toStringRef(const BlobCipherEncryptHeaderRef& headerRef) {
+		return BinaryWriter::toValue(headerRef, IncludeVersion());
+	}
+
+	// Routine computes EncryptHeaderSize based on input params.
+	static uint32_t getHeaderSize(const int flagVersion,
+	                              const int algoHeaderVersion,
+	                              const EncryptCipherMode cipherMode,
+	                              const EncryptAuthTokenMode authMode,
+	                              const EncryptAuthTokenAlgo authAlgo);
+
+	// Assists in extracting necessary information from the headerRef string (on-disk read)
+	static ParsedEncryptHeaderDetails getCipherDetailsAndIv(const BlobCipherEncryptHeaderRef&);
+
+	template <class Ar>
+	void serialize(Ar& ar);
```

Comment:
To enable code outside of BlobCipher to use the serializer, it has to be define (implement) in the .h file.

Reply:
I modified BlobGranuleFile encryption code and it worked fine?

Reply:
I didn't look at the blob granule PR, but if it is using the fromStringRef helper then it doesn't matter. But if some other code want to nested the header in an other struct, or, for example, use the `BinaryWriter <<` operator, then it will matter.

Reply:
I see; yeah it uses StringRef header serialized value, I think it makes sense to keep the implementation inline to the class, it was kinda a decent long method, so pulled it out. 

Will `inline` the method for now.

Reply:
sure. the method will become short while I do the `std::variant` serialization refactor, and I can move it with that PR.